### PR TITLE
Implement MouseEvent's x/y and offsetX/offsetY attributes

### LINKED ISSUE #24248: Missing MouseEvent attributes defined by CSSOM View
https://drafts.csswg.org/cssom-view/#extensions-to-the-mouseevent-interface

In particular, the pageX/pageY attributes are used by three.js code in http://rawcdn.githack.com/mrdoob/three.js/r105/examples/js/controls/FirstPersonControls.js . This is likely to be a source of surprising web compat issues!

This will likely have an impact on #20931.

### PR DESCRIPTION
**Pull request by tigleym**
Implement MouseEvent's x/y and offsetX/offsetY attributes

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #24248 (GitHub issue number if applicable)

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: tests/wpt/metadata/css/cssom-view/mouseEvent.html.ini [hunk 1]
```diff
@@ -1,4 +1 @@
 [mouseEvent.html]
-  [MouseEvent's x and y must be equal to clientX and clientY.]
-    expected: FAIL
-
```
Comment (atouchet) [inline tests/wpt/metadata/css/cssom-view/mouseEvent.html.ini]: If there are no more failing tests you can delete the file.

File: components/script/dom/mouseevent.rs [hunk 2]
```diff
@@ -192,6 +203,56 @@ impl MouseEventMethods for MouseEvent {
...
+        self.client_y.get()
+    }
+
+    // https://drafts.csswg.org/cssom-view/#dom-mouseevent-offsetx
+    fn OffsetX(&self) -> i32 {
+        let event = self.upcast::<Event>();
+        if event.dispatching() {
+            match event.GetTarget() {
+                Some(target) => {
+                    if let Some(element) = target.downcast::<Element>() {
```
Comment (jdm) [inline components/script/dom/mouseevent.rs]: If we downcast to Node instead then we can call `Node::client_rect`. This means we should get expected behaviour for mouse events on non-Elements like text nodes, for example.

File: components/script/dom/mouseevent.rs [hunk 3]
```diff
@@ -192,6 +203,56 @@ impl MouseEventMethods for MouseEvent {
...
+        }
+    }
+
+    // https://drafts.csswg.org/cssom-view/#dom-mouseevent-offsety
+    fn OffsetY(&self) -> i32 {
+        let event = self.upcast::<Event>();
+        if event.dispatching() {
+            match event.GetTarget() {
+                Some(target) => {
+                    if let Some(element) = target.downcast::<Element>() {
```
Comment (jdm) [inline components/script/dom/mouseevent.rs]: Same as previous comment.

File: components/script/dom/domrect.rs [hunk 4]
```diff
@@ -32,6 +32,14 @@ impl DOMRect {
         )
     }
 
+    pub fn left(&self) -> f64 {
+        self.rect.Left()
+    }
+
+    pub fn top(&self) -> f64 {
```
Comment (jdm) [inline components/script/dom/domrect.rs]: We should be able to remove these based on my other comments.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
This looks good, it just needs a final `./mach update-manifest` to reflect the changes to the WPT test.

**Comment by tigleym** [comment]
> This looks good, it just needs a final `./mach update-manifest` to reflect the changes to the WPT test.

Sounds good! I'll get this updated today.

**Pull request merged**